lib/SecretsCrypto/secrets.c

```diff
@@ -0,0 +1,1000 @@
+Sailfish_Secrets_Callback_Data_new(
+		Sailfish_Secrets_SecretManager_result_callback rc,
+		Sailfish_Secrets_SecretManager_secret_result_callback src,
+		void *cc)
+{
+	struct Sailfish_Secrets_Callback_Data *cbd =
+	    (struct Sailfish_Secrets_Callback_Data *)
+	    malloc(sizeof(struct Sailfish_Secrets_Callback_Data));
+
+	cbd->result_callback = rc;
+	cbd->secret_result_callback = src;
+	cbd->callback_context = cc;
+	cbd->refcount = 1;
+
+	return cbd;
+}
+
+void Sailfish_Secrets_Callback_Data_ref(
+		struct Sailfish_Secrets_Callback_Data *cbd)
+{
+	cbd->refcount = cbd->refcount + 1;
+}
+
+void Sailfish_Secrets_Callback_Data_unref(
+		struct Sailfish_Secrets_Callback_Data *cbd)
+{
+	cbd->refcount = cbd->refcount - 1;
+	if (cbd->refcount == 0)
+		free(cbd);
+}
+
+/******************************* Internal Daemon Connection *****************/
+
+static struct Sailfish_Secrets_DBus_Connection {
+	GDBusProxy *discoveryProxy;
+	char *p2pAddr;
+	GDBusConnection *p2pBus;
+	GDBusProxy *secretsProxy;
+	int busy;
+} daemon_connection = {
+	.discoveryProxy = NULL,
+	.p2pAddr = NULL,
+	.p2pBus = NULL,
+	.secretsProxy = NULL,
+	.busy = 0
+};
+
+void Sailfish_Secrets_proxyReady(
+		GObject *source_object,
+		GAsyncResult *res,
+		gpointer user_data)
+{
+	GDBusProxy *proxy = g_dbus_proxy_new_finish(res, NULL);
+	(void)source_object;
+	if (proxy) {
+		/* Success, we're connected! */
+		struct Sailfish_Secrets_Result *result =
+		    Sailfish_Secrets_Result_new
+		    (Sailfish_Secrets_Result_Succeeded, 0, "");
+		struct Sailfish_Secrets_Callback_Data *cbd =
+		    (struct Sailfish_Secrets_Callback_Data *)user_data;
+		daemon_connection.secretsProxy = proxy;
+		daemon_connection.busy = 0;
+		cbd->result_callback(cbd->callback_context, result);
+		Sailfish_Secrets_Result_unref(result);
+		Sailfish_Secrets_Callback_Data_unref(cbd);
+	} else {
+		struct Sailfish_Secrets_Result *result =
+		    Sailfish_Secrets_Result_new(
+				    Sailfish_Secrets_Result_Failed,
+				    5,
+				    "Unable to create secrets interface");
+		struct Sailfish_Secrets_Callback_Data *cbd =
+		    (struct Sailfish_Secrets_Callback_Data *)user_data;
+		g_dbus_connection_close_sync(daemon_connection.p2pBus, NULL,
+					     NULL);
+		g_object_unref(daemon_connection.p2pBus);
+		daemon_connection.p2pBus = NULL;
+		daemon_connection.busy = 0;
+		cbd->result_callback(cbd->callback_context, result);
+		Sailfish_Secrets_Result_unref(result);
+		Sailfish_Secrets_Callback_Data_unref(cbd);
+	}
+}
+
+void Sailfish_Secrets_connectionReady(
+		GObject *source_object,
+		GAsyncResult *res,
+		gpointer user_data)
+{
+	daemon_connection.p2pBus =
+	    g_dbus_connection_new_for_address_finish(res, NULL);
+	(void)source_object;
+	if (daemon_connection.p2pBus) {
+		free(daemon_connection.p2pAddr);
+		daemon_connection.p2pAddr = NULL;
+		g_dbus_proxy_new(
+		    daemon_connection.p2pBus,
+		    G_DBUS_PROXY_FLAGS_NONE,
+		    NULL,
+		    NULL, /* bus name */
+		    "/Sailfish/Secrets",
```

Comment:
Also. I would create a header file to store all string and numeric (returned values, any parameters, bus names, etc) constants in one place. It's convenient if one need to modify it.

Reply:
Ok